File: Gstreamer-TFlite/src/tflite_object_detection.c
/*
* Copyright (c) 2020 Qualcomm Innovation Center, Inc.  All Rights Reserved.
*
* SPDX-License-Identifier: BSD-3-Clause-Clear
*/
#include <gst/gst.h>

int main(int argc, char *argv[]) {

    GstElement *pipeline, *source, *sink, *capsfilter;
    GstElement *qtimletflite, *queue, *qtioverlay;
    GstBus *bus;
    GstMessage *msg;
    GstStateChangeReturn ret;

    /* Initialize GStreamer */
    gst_init (&argc, &argv);

    /* Create the elements */
    source          =   gst_element_factory_make ("qtiqmmfsrc", "source");
    capsfilter      =   gst_element_factory_make ("capsfilter", "caps-filter");
    qtimletflite    =   gst_element_factory_make ("qtimletflite", "qti-mletflite");
    queue           =   gst_element_factory_make ("queue", "queue");
    qtioverlay      =   gst_element_factory_make ("qtioverlay", "qti-overlay");
    sink            =   gst_element_factory_make ("waylandsink", "sink");

    /* Create the empty pipeline */
    pipeline = gst_pipeline_new ("test-pipeline");

    if (!pipeline || !source || !sink || !capsfilter || !qtimletflite || !queue || !qtioverlay) {
        g_printerr ("Not all elements could be created.\n");
        return -1;
    }

    /* Build the pipeline */
    gst_bin_add_many (GST_BIN (pipeline), source, capsfilter, qtimletflite, queue, qtioverlay, sink, NULL);
    if (gst_element_link_many (source, capsfilter, qtimletflite, queue, qtioverlay, sink, NULL) != TRUE) {
        g_printerr ("Elements could not be linked.\n");
        gst_object_unref (pipeline);
        return -1;
    }

    /* Modify the properties for caps-fitler */
    g_object_set (capsfilter, "caps",
            gst_caps_from_string ("video/x-raw,format=NV12,width=1280,height=720,framerate=30/1,camera=0"), NULL);

    /* Modify the properties for qtimle-tflite */
    g_object_set (qtimletflite, "config", "/data/misc/camera/mle_tflite.config", 
        "model", "/data/misc/camera/detect.tflite", "labels", 
        "/data/misc/camera/labelmap.txt", "postprocessing", "detection", NULL);

    /* Modify the properties for sink */
    g_object_set (sink, "fullscreen", TRUE, "async", TRUE, "enable-last-sample", FALSE, NULL);

    /* Start playing */
    ret = gst_element_set_state (pipeline, GST_STATE_PLAYING);
    if (ret == GST_STATE_CHANGE_FAILURE) {
        g_printerr ("Unable to set the pipeline to the playing state.\n");
        gst_object_unref (pipeline);
        return -1;
    }

    /* Wait until error or EOS */
    bus = gst_element_get_bus (pipeline);
    msg = gst_bus_timed_pop_filtered (bus, GST_CLOCK_TIME_NONE, GST_MESSAGE_ERROR | GST_MESSAGE_EOS);

    /* Parse message */
    if (msg != NULL) {
        GError *err;
        gchar *debug_info;

        switch (GST_MESSAGE_TYPE(msg)) {
        case GST_MESSAGE_ERROR:
            gst_message_parse_error (msg, &err, &debug_info);
            g_printerr ("Error received from element %s: %s\n", GST_OBJECT_NAME (msg->src), err->message);
            g_printerr ("Debugging information: %s\n", debug_info ? debug_info : "none");
            g_clear_error (&err);
            g_free (debug_info);
            break;
        case GST_MESSAGE_EOS:
            g_print ("End-Of-Stream reached.\n");
            break;
        default:
            /* We should not reach here because we only asked for ERRORs and EOS */
            g_printerr ("Unexpected message received.\n");
            break;
        }
        gst_message_unref (msg);
    }

    /* Free resources */
    gst_object_unref (bus);
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref (pipeline);
    return 0;
}

File: gst-python-samples/Gstreamer-Applications/gst_tflite/src/tflite_object_detection.py
#!/usr/bin/env python3
import sys
import gi
import logging


gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")

from gi.repository import Gst, GLib

logging.basicConfig(level=logging.DEBUG,
                    format="[%(name)s] [%(levelname)8s] - %(message)s")
logger = logging.getLogger(__name__)

if __name__ == "__main__":

    # initialize GStreamer
    Gst.init(sys.argv)

    loop = GLib.MainLoop()

    # create element
    source = Gst.ElementFactory.make("qtiqmmfsrc", "source")
    capsfilter = Gst.ElementFactory.make("capsfilter", "caps-filter")
    qtimletflite = Gst.ElementFactory.make("qtimletflite", "qti-mletflite")
    queue = Gst.ElementFactory.make("queue", "queue")
    qtioverlay = Gst.ElementFactory.make("qtioverlay", "qti-overlay")
    sink = Gst.ElementFactory.make("waylandsink", "sink")

    # creating empty pipeline
    pipeline = Gst.Pipeline.new("tflite-object-detection")

    if not source or not capsfilter or not qtimletflite or not queue or not qtioverlay or not sink:
        print("Create element failed")
        sys.exit(1)

    # Build the pipeline
    pipeline.add(source)
    pipeline.add(capsfilter)
    pipeline.add(qtimletflite)
    pipeline.add(queue)
    pipeline.add(qtioverlay)
    pipeline.add(sink)

    if not Gst.Element.link(source, capsfilter):
        print("Source & Capsfilter can not be linked")
        sys.exit(1)
    if not Gst.Element.link(capsfilter, qtimletflite):
        print("Capsfilter & qtitflite can not be linked")
        sys.exit(1)
    if not Gst.Element.link(qtimletflite, queue):
        print("qtitflite & queue can not be linked")
        sys.exit(1)
    if not Gst.Element.link(queue, qtioverlay):
        print("Queue & Qtioverlay could not be linked")
        sys.exit(1)
    if not Gst.Element.link(qtioverlay, sink):
        print("Qtioverlay & Sink could not be linked")
        sys.exit(1)

    # Modify the properties for caps-fitler
    capsfilter.set_property("caps", Gst.caps_from_string(
        "video/x-raw,format=NV12,width=1280,height=720,framerate=30/1,camera=0"))

    # Modify the properties for qtimle-tflite
    qtimletflite.set_property("config", "/data/misc/camera/mle_tflite.config")
    qtimletflite.set_property("model", "/data/misc/camera/detect.tflite")
    qtimletflite.set_property("labels", "/data/misc/camera/labelmap.txt")
    qtimletflite.set_property("postprocessing", "detection")

    # Modify the properties for sink
    sink.set_property("fullscreen", "true")
    sink.set_property("async", "true")
    sink.set_property("enable-last-sample", "true")

try:
    # Start playing
    pipeline.set_state(Gst.State.PLAYING)
    print("Object Detection Started..")
    print('Press Ctrl+C to stop')
    loop.run()
except KeyboardInterrupt:
    print("Stopping Object Detection")
    pipeline.set_state(Gst.State.NULL)
    sys.exit(0)
